packages/react-components/src/components/chart/hooks/useEvents.ts

```diff
@@ -0,0 +1,135 @@
+        const graphicIndex = graphic.findIndex(
+          (g) => g.children[TREND_CURSOR_CLOSE_GRAPHIC_INDEX].id === e?.target?.id
+        );
+
+        // delete tc
+        if (graphicIndex !== -1) {
+          if (isInSyncMode) {
+            deleteTrendCursorsInSyncState({ groupId: groupId ?? '', tcId: graphic[graphicIndex].id as string });
+          } else {
+            graphic[graphicIndex].$action = 'remove';
+            graphic[graphicIndex].children = []; // Echarts will throw error if children are not empty
+            chart?.setOption({ graphic });
+            graphic.splice(graphicIndex, 1);
+            setGraphic([...graphic]);
+          }
+        }
+
+        // add TC
+        if (isInCursorAddMode && graphic.length < MAX_TREND_CURSORS) {
+          if (isInSyncMode) {
+            const timestampInMs = calculateTimeStamp(e?.offsetX ?? 0, size.width, viewport);
+            addTrendCursorsToSyncState({
+              groupId: groupId ?? '',
+              tcId: `trendCursor-${uuid()}`,
+              timestamp: timestampInMs,
+              tcHeaderColorIndex: trendCursorStaticIndex++,
+            });
+          } else {
+            const newTc = getNewTrendCursor({
+              e,
+              size,
+              tcHeaderColorIndex: trendCursorStaticIndex++,
+              series,
+              yMin,
+              yMax,
+              viewport,
+            });
+
+            setGraphic([...graphic, newTc]);
+          }
+        }
+      });
+
+      // this is the ondrag handler of the trend cursor
+      chart?.getZr().on('drag', (event) => {
+        if (event.target.id.toString().startsWith('line')) {
+          const graphicIndex = graphic.findIndex((g) => g.children[0].id === event.target.id);
+          const timeInMs = calculateTimeStamp(event.offsetX ?? 0, size.width, viewport);
+          if (isInSyncMode) {
+            updateTrendCursorsInSyncState({
+              groupId: groupId ?? '',
+              timestamp: timeInMs,
+              tcId: graphic[graphicIndex].id as string,
+            });
+          } else {
+            // update current TC
+            graphic[graphicIndex] = ondragUpdateTrendCursor({
+              graphic: graphic[graphicIndex],
+              posX: event.offsetX ?? 0,
+              timeInMs,
+              size,
+              series,
+              yMax,
+              yMin,
+            });
+            // update component state
+            setGraphic([...graphic]);
+          }
+        }
+      });
+    }
+
+    return () => {
+      if (ref.current !== null) {
+        chart?.getZr().off('click');
+        // passing the function so that it will NOT remove all the mouseover behaviour, namely default tooltip
+        chart?.getZr().off('mouseover', () => mouseoverHandler(isInCursorAddMode, chart));
+        chart?.getZr().off('drag');
+      }
+    };
+  }, [ref, graphic, size, isInCursorAddMode, setGraphic, viewport, series, yMin, yMax, isInSyncMode]);
```

Comment:
Something tells me there may be a better way to do this than in an effect. Do we need to run this code every time any of these props change?

Reply:
yes, so we can just add the event handlers w/o the use of useEffect, but how would you unsubscribe? and also w/o a empty dependency list... it will run only once, but how to make sure we get the latest value ?

Reply:
following the pattern suggested here in this link 
https://stackoverflow.com/questions/55565444/how-to-register-event-with-useeffect-hooks